### DISCUSSION TITLE
row and even column based permissions

### DISCUSSION THREAD

**Discussion by tcurdt** [pocketbase/pocketbase#2341] [Q&A] [answered]
AFAIU from the docs there are only row level permissions for now.
It would require to split the collection for finer control.

> Relation (Ownership)
> Let's say that you have 2 collections - "posts" base collection and "users" auth collection. In your "posts" collection you can create "author" relation field pointing to the "users" collection. To allow access to only the "author" of the record(s), you could use a rule like: @request.auth.id != "" && author = @request.auth.id
> Nested relation fields look ups are also supported, eg: someRelField.anotherRelField.author = @request.auth.id

So if a user wants to update its avatar but should not be allowed to update their subscription the suggested way would be to move the avatar and/or the subscription into its own collection. And then control access on the collection with:

    @request.auth.id ?= user_id

Or does it have to be `@request.auth.id != "" && user_id = @request.auth.id`.
And why is it `=` and not `==` in the example in the docs?

And in the future [there will be a](https://github.com/pocketbase/pocketbase/pull/759#issuecomment-1435731789) `view` feature?

**Comment by ganigeorgiev** [comment]
> AFAIU from the docs there are only row level permissions for now.

I'm not entirely sure that I understand what do you mean. The possible permission controls are outlined in https://pocketbase.io/docs/collections/#auth-collection.

---

> So if a user wants to update its avatar but should not be allowed to update their subscription the suggested way would be to move the avatar and/or the subscription into its own collection.

Yes, that's one way to handle it. Alternatively, you can also check the submitted value within the API rule, for example with the [`:isset` modifier](https://pocketbase.io/docs/api-rules-and-filters/#isset-modifier).

---

> And why is it = and not == in the example in the docs?

Because there are no assignments in the rules and there is no need to distinguish `=` from `==`/`===` and it is similar to the `WHERE` SQL syntax.

---

> And in the future https://github.com/pocketbase/pocketbase/pull/759#issuecomment-1435731789 view feature?

There is already a View collection with PocketBase v0.13+.

**Comment by tcurdt** [comment]
>> AFAIU from the docs there are only row level permissions for now.
> I'm not entirely sure that I understand what do you mean.

When a user is owner of a post, they can update their full row from the post collection.
But you cannot control access to certain fields of that row.
That's why moving to another collection is necessary.
Right?

But wouldn't that be a security issue? 🤔

So if the user can change the full row, could they also change the relation?
In the example from the docs they could change the user relation.
And their post would become the post of another account.
No?

Cool, the view stuff has already landed 👍

**Reply by ganigeorgiev** [reply, marked as answer]
> That's why moving to another collection is necessary.

Not exactly. You can check the submitted value in the API rule and prevent user overwriting it. Here is the example from the docs using the `:isset` modifier

```js
// disallows changing a "role" named field
@request.data.role:isset = false

// alternatively you can also check it against an existing value
// or for example the current authenticated user (assuming the field is marked nonempty)
@request.data.author = @request.auth.id
```

I'm not sure how to explain it better. If you haven't already, you could check https://pocketbase.io/docs/api-rules-and-filters/#api-rules. If you have a suggestion for better and more clear explanation feel free to file a PR in https://github.com/pocketbase/site or elaborate what part of the docs are not clear and I'll try to improve them.

**Comment by tcurdt** [comment]
Aaaah. I didn't make that connection.
Thanks for the help!
I'll see if I can come up with a PR.

Last question :)
Is there a reason this

`@request.auth.id != "" && author = @request.auth.id`

couldn't be written as such?

`author = @request.auth.id`

If the author is `non-empty` I don't see the need for another non-empty check.

**Reply by ganigeorgiev** [reply]
> If the author is non-empty I don't see the need for another non-empty check.

Yes, the extra empty string check is not needed if the `author` field is `required`, aka. marked "Nonempty", since you'll get a validation error for the field.

**Comment by tcurdt** [comment]
Thanks!
